lib/pure/strutils.nim

```diff
@@ -2860,6 +2860,113 @@ proc isNilOrWhitespace*(s: string): bool {.noSideEffect, procvar, rtl,
   ## Alias for isEmptyOrWhitespace
   result = isEmptyOrWhitespace(s)
 
+
+since (1, 1):
+  template toLowerAsciiInPlace*(c: var char) =
+    ## Returns the lower case version of character ``c``.
+    ##
+    ## This works only for the letters ``A-Z``. See `unicode.toLower
+    ## <unicode.html#toLower,Rune>`_ for a version that works for any Unicode
+    ## character.
+    ##
+    ## See also:
+    ## * `isLowerAscii proc<#isLowerAscii,char>`_
+    ## * `toLowerAscii proc<#toLowerAscii,string>`_ for converting a string
+    runnableExamples:
+      var character = 'F'
+      toLowerAsciiInPlace(character)
+      doAssert character == 'f'
+      var chara = 'A'
+      toLowerAsciiInPlace(chara)
+      doAssert chara == 'a'
+    if c in {'A'..'Z'}: c = chr(ord(c) + (ord('a') - ord('A')))
+
+
+  template toUpperAsciiInPlace*(c: var char) =
+    ## Converts character `c` into upper case.
+    ##
+    ## This works only for the letters ``A-Z``.  See `unicode.toUpper
+    ## <unicode.html#toUpper,Rune>`_ for a version that works for any Unicode
+    ## character.
+    ##
+    ## See also:
+    ## * `isLowerAscii proc<#isLowerAscii,char>`_
+    ## * `toUpperAscii proc<#toUpperAscii,string>`_ for converting a string
+    ## * `capitalizeAscii proc<#capitalizeAscii,string>`_
+    runnableExamples:
+      var character = 'f'
+      toUpperAsciiInPlace(character)
+      doAssert character == 'F'
+      var chara = 'z'
+      toUpperAsciiInPlace(chara)
+      doAssert chara == 'Z'
+    if c in {'a'..'z'}: c = chr(ord(c) - (ord('a') - ord('A')))
```

Comment:
But... why? Is it that slow to return a char?

Reply:
Yeah this makes no sense.

Reply:
Try for yourself:

```nim
import times
from strutils import toLowerAscii

template toLowerAscii2(c: var char) =
  if c in {'A'..'Z'}: c = chr(ord(c) + (ord('a') - ord('A')))

echo "strutils.toLowerAscii"
var char1: char
let t0 = now()
let t0a = cpuTime()
for _ in 0..99:
  char1 = 'A'
  discard strutils.toLowerAscii(char1)
echo now() - t0, '\t', cpuTime() - t0a

echo "toLowerAscii2"
var char0: char
let t1 = now()
let t1a = cpuTime()
for _ in 0..99:
  char0 = 'A'
  toLowerAscii2(char0)
echo now() - t1, '\t', cpuTime() - t1a
```

---

```console
$ nim c -r example2.nim
strutils.toLowerAscii
55 microseconds and 883 nanoseconds     4.047999999999899e-05
toLowerAscii2
9 microseconds and 76 nanoseconds      1.158199999999883e-05

$ nim c -r -d:release -d:danger example2.nim
strutils.toLowerAscii
32 microseconds and 9 nanoseconds       4.101599999999962e-05
toLowerAscii2
8 microseconds and 804 nanoseconds     1.222200000000023e-05
```

:man_shrugging:

Reply:
So I tweaked the test a bit:

```nim
import std/monotimes, times, strutils
proc toLowerAscii1(c: char): char {.inline.} =
  # same as stdlib but with {.inline.}
  if c in {'A'..'Z'}:
    result = chr(ord(c) + (ord('a') - ord('A')))
  else:
    result = c
proc toLowerAscii2(c: var char) {.inline.} =
  if c in {'A'..'Z'}: c = chr(ord(c) + (ord('a') - ord('A')))

template bench(body: untyped): untyped =
  block:
    let start = getMonoTime()
    var ch {.inject.}: char
    for _ in 0..99:
      ch = 'A'
      body
    echo getMonoTime() - start

echo "strutils.toLowerAscii"
bench: ch = toLowerAscii(ch)

echo "toLowerAscii1"
bench: ch = toLowerAscii1(ch)

echo "toLowerAscii2"
bench: toLowerAscii2(ch)
```

And the results:

```
$ nim c -r test.nim
strutils.toLowerAscii
2 microseconds and 170 nanoseconds
toLowerAscii1
1 microsecond and 800 nanoseconds
toLowerAscii2
1 microsecond and 731 nanoseconds

$ nim c -d:danger -r test.nim
strutils.toLowerAscii
330 nanoseconds
toLowerAscii1
20 nanoseconds
toLowerAscii2
20 nanoseconds
```

It seems to me that the only overhead here is the function call overhead, you can easily fix this by adding `{.inline.}` into the stdlib version.

Reply:
Thats `330` Vs `20`, that confirms is faster. :man_shrugging: 

I think that the point is that difference accumulates in the loop, 
I dont care too much about the `char` functions, 
but the `string` function the difference is bigger.

I got more difference (Microseconds Vs Nanoseconds, but I believe you).
What do we do?, add `{.inline.}` to all the `strutils` ones?, close this PR?. :thinking:

Reply:
> Thats `330` Vs `20`, that confirms is faster.

Well of course there will be a difference when you `{.inline.}` vs calling functions. The point here is `var char` doesn't really improve anything, so we don't need to add new APIs for that, just adding `{.inline.}` strategically is enough.

> I dont care too much about the `char` functions

You may not care about it, but it does matter when a new API is added to the stdlib. We should be conservative about adding new APIs, since once added we will be stuck with it for the rest of the 1.x series.

With that said, I don't have any objections on adding the in-place versions for the `string` variant, as there are actual speed up there.

> add `{.inline.}` to all the strutils ones?

We should only target simple procs (ie. < 3 lines). Better would be to make the compiler able to perform this optimization on it's own, though that's a lot harder.

Reply:
Done. 
Only `string` proc is public.